src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentRenderModeDirectivePass.cs

```diff
@@ -0,0 +1,108 @@
+{
+    protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+    {
+        var @namespace = documentNode.FindPrimaryNamespace();
+        var @class = documentNode.FindPrimaryClass();
+        if (@namespace == null || @class == null)
+        {
+            return;
+        }
+
+        var directives = documentNode.FindDirectiveReferences(ComponentRenderModeDirective.Directive);
+        if (directives.Count == 0)
+        {
+            return;
+        }
+
+        // We don't need to worry about duplicate attributes as we have already replaced any multiples with MalformedDirective
+        Debug.Assert(directives.Count == 1);
+
+        var token = ((DirectiveIntermediateNode)directives[0].Node).Tokens.FirstOrDefault();
+        if (token == null)
+        {
+            return;
+        }
+
+        // generate the inner attribute class
+        // PROTOTYPE: extract out to consts
+        var classDecl = new ClassDeclarationIntermediateNode()
+        {
+            ClassName = "PrivateComponentRenderModeAttribute",
+            BaseType = "global::Microsoft.AspNetCore.Components.RenderModeAttribute",
+        };
+        classDecl.Modifiers.Add("file");
+        classDecl.Modifiers.Add("sealed");
+        classDecl.Children.Add(new CSharpCodeIntermediateNode()
+        {
+            Children =
+            {
+                new IntermediateToken()
+                {
+                    Kind = TokenKind.CSharp,
+                    Content = "private static IComponentRenderMode ModeImpl => "
+                },
+                new CSharpCodeIntermediateNode()
+                {
+                    Source = token.Source,
+                    Children =
+                    {
+                        new IntermediateToken()
+                        {
+                            Kind = TokenKind.CSharp,
+                            Content = token.Content
+                        }
+                    }
+                },
+                new IntermediateToken()
+                {
+                    Source = token.Source,
+                    Kind = TokenKind.CSharp,
+                    Content = ";"
+                }
+            }
+        });
+        classDecl.Children.Add(new CSharpCodeIntermediateNode()
+        {
+            Children =
+            {
+                new IntermediateToken()
+                {
+                    Kind = TokenKind.CSharp,
+                    Content = "public override IComponentRenderMode Mode => ModeImpl;"
+                }
+            }
+        });
+
+        // generate the attribute usage on top of the class
+        var attributeNode = new CSharpCodeIntermediateNode();
+        attributeNode.Children.Add(new IntermediateToken()
+        {
+            Kind = TokenKind.CSharp,
+            Content = $"[global::{@namespace.Content}.PrivateComponentRenderModeAttribute]",
```

Comment:
Just chuckling a bit at using `global::` to get to a `file` type but agree this is the way to do it.